ampd/src/event_processor.rs

```diff
@@ -541,37 +637,288 @@ mod tests {
+        let mut broadcaster = MockBroadcasterClient::new();
+        broadcaster
+            .expect_broadcast()
+            .times(events.len())
+            .returning(|_| Ok(()));
+
+        let (server, monitoring_client) = monitoring::Server::new(None).unwrap();
+        let cancel = CancellationToken::new();
+        let server_handle = tokio::spawn(server.run(cancel.clone()));
+
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let result_with_timeout = timeout(
+            Duration::from_secs(3),
+            consume_events(
+                handler,
+                broadcaster,
+                stream::iter(events),
+                event_config,
+                cancel.clone(),
+                monitoring_client.clone(),
+                setup_handler_info(true),
+            ),
+        )
+        .await;
+
+        assert!(result_with_timeout.is_ok());
+        assert!(result_with_timeout.unwrap().is_ok());
+
+        cancel.cancel();
+        let _ = server_handle.await;
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn record_vote_metrics_successfully_when_encounters_handler_failure() {
+        let events: Vec<Result<Event, event_processor::Error>> =
+            vec![Ok(Event::BlockEnd(0_u32.into()))];
+
+        let mut handler = MockEventHandler::new();
+        handler
+            .expect_handle()
+            .times(3)
+            .returning(|_| Err(report!(EventHandlerError::Failed)));
+
+        let broadcaster = MockBroadcasterClient::new();
+        let event_config = setup_event_config(
+            Duration::from_secs(1),
+            Duration::from_secs(1000),
+            Duration::from_secs(1),
+        );
+
+        let bind_addr = localhost_with_random_port();
+        let (server, monitoring_client) = monitoring::Server::new(bind_addr).unwrap();
+        let cancel_token = CancellationToken::new();
+        let server_handle = tokio::spawn(server.run(cancel_token.clone()));
+
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let result_with_timeout = timeout(
+            Duration::from_secs(3),
+            consume_events(
+                handler,
+                broadcaster,
+                stream::iter(events),
+                event_config,
+                cancel_token.clone(),
+                monitoring_client.clone(),
+                setup_handler_info(true),
+            ),
+        )
+        .await;
+
+        assert!(result_with_timeout.is_ok());
+        assert!(result_with_timeout.unwrap().is_ok());
+
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let metrics_url = create_endpoint_url(bind_addr, "metrics");
+
+        let response = reqwest::get(metrics_url).await.unwrap();
+        let metrics_text = response.text().await.unwrap();
+
+        let sorted_metrics = sort_metrics_text(metrics_text);
+        goldie::assert!(sorted_metrics);
+
+        cancel_token.cancel();
+        let _ = server_handle.await;
+    }
```

Comment:
this is testing stuff that is already tested in the monitoring module and that this shouldn't have to know about.
Instead, make a test server available in the monitoring module that gives you direct access to messages sent by the client. Then you can test the output of the handler directly without having to go through all the monitoring setup that the handler shouldn't be concerned with